burn-core/src/optim/rmsprop.rs

```diff
@@ -0,0 +1,500 @@
+                Some(state) => state
+                    .buf
+                    .clone()
+                    .mul_scalar(self.momentum)
+                    .add(grad.clone()),
+                _ => grad.clone(),
+            };
+            (buf.clone(), centered_state, RMSPropMomentumState { buf })
+        } else {
+            match momentum_state {
+                Some(state) => (grad.clone(), centered_state, state),
+                _ => (
+                    grad.clone(),
+                    centered_state,
+                    RMSPropMomentumState {
+                        buf: Tensor::zeros(grad.shape()),
+                    },
+                ),
+            }
+        }
+    }
+}
+
+/// [RMSPropMomentumState](RMSPropMomentumState) is to store and pass optimizer step params.
+#[derive(Record, Clone, new)]
+pub struct RMSPropMomentumState<B: Backend, const D: usize> {
+    buf: Tensor<B, D>,
+}
+
+impl<B: Backend, const D: usize> RMSPropMomentumState<B, D> {
+    fn to_device(mut self, device: &B::Device) -> Self {
+        self.buf = self.buf.to_device(device);
+        self
+    }
+}
+
+impl<B: Backend> SimpleOptimizer<B> for RMSProp<B> {
+    type State<const D: usize> = RMSPropState<B, D>;
+
+    fn step<const D: usize>(
+        &self,
+        lr: LearningRate,
+        mut tensor: Tensor<B, D>,
+        mut grad: Tensor<B, D>,
+        state: Option<Self::State<D>>,
+    ) -> (Tensor<B, D>, Option<Self::State<D>>) {
+        // fetch state for params
+        let mut state_square_avg = None;
+        let mut state_centered = None;
+        let mut state_momentum = None;
+        if let Some(state) = state {
+            state_square_avg = Some(state.square_avg);
+            state_centered = Some(state.centered);
+            state_momentum = Some(state.momentum);
+        }
+
+        // weight_decay transform
+        if let Some(weight_decay) = &self.weight_decay {
+            let (grad_out, tensor_out) = weight_decay.transform_temp_fix(grad, tensor);
+            grad = grad_out;
+            tensor = tensor_out;
+        }
+
+        // square_avg transform
+        let (mut grad, state_square_avg) =
+            SquareAvgState::transform(self.alpha, grad, state_square_avg);
+
+        // centered trnsform
+        let (grad_out, state_square_avg, state_centered) = CenteredState::transform(
+            self.alpha,
+            self.centered,
+            grad,
+            state_square_avg,
+            state_centered,
+        );
+        grad = grad_out;
+
+        // momentum transform
+        let (grad, state_centered, state_momentum) =
+            self.momentum
+                .transform(grad, state_centered, state_momentum);
+
+        // transition state
+        let state = RMSPropState::new(state_square_avg, state_centered, state_momentum);
+
+        // tensor param transform
+        let delta = grad.mul_scalar(lr);
+        // (tensor - delta, Some(state))
+        let tmp = tensor - delta;
```

Comment:
`tmp` can be `update` ?

Reply:
sorry forget to remove the var for debugging 😅